src/librustc/ty/context.rs

```diff
@@ -1072,6 +1072,23 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
         local as usize == global as usize
     }
 
+    /// Returns true if this function must conform to `min_const_fn`
+    pub fn is_min_const_fn(self, def_id: DefId) -> bool {
+        match self.lookup_stability(def_id) {
+            // stable functions with unstable const fn don't need to conform
+            Some(&attr::Stability { const_stability: Some(_), .. }) => false,
+            // unstable functions don't need to conform
+            Some(&attr::Stability { ref level, .. }) if level.is_unstable() => false,
+            _ => {
+                // libstd stability has been checked above, so everything else must
+                // be `min_const_fn`
+                // BUT: users enabling the `const_fn` can do what they want
+                // just the libstd has to be nice
+                self.features().staged_api || !self.sess.features_untracked().const_fn
```

Comment:
Would be good to have a test which tests that libstd/liballoc/libcore (or a fake crate that is allowed stability attributes) functions can't have stabilized `const fn`s which don't conform to `min_const_fn`